File: cadwiki-nuget/cadwiki.DllReloader/AutoCAD/AutoCADAppDomainDllReloader.vb
Option Strict On
Option Infer Off
Option Explicit On

Imports System.Reflection
Imports Autodesk.AutoCAD.ApplicationServices
Imports Autodesk.AutoCAD.Runtime
Imports System.IO
Imports System.Globalization
Imports Autodesk.AutoCAD.EditorInput

Namespace AutoCAD
    Public Class AutoCADAppDomainDllReloader
        Inherits AutodeskAppDomainReloader

        Private _document As Document
        Private _tempFolder As String

        Public Shadows Function GetLogMode() As LogMode
            Return DependencyValues.LogMode
        End Function

        Public Sub Configure(currentIExtensionAppAssembly As Assembly)
            Try
                Log("---------------------------------------------")
                Log("---------------------------------------------")
                Log("Configure started.")
                ReadDependecyValuesFromIni()
                ' If Terminated = True
                ' And Versions don't match
                ' this is the first Initalize call from any IExtensionApplication in this AutoCAD session
                Dim wasLastReloaderStateTerminated As Boolean = GetTerminatedFlag()
                If wasLastReloaderStateTerminated = True Then
                    'Not NetReloader.GetVersion().Equals(iExtensionAppVersion) Then
                    SetReloadCount(0)
                    SetInitialValues(currentIExtensionAppAssembly)
                End If

                If String.IsNullOrEmpty(GetIExtensionApplicationClassName()) Then
                    SetIExtensionApplicationClassNameFromAssembly(currentIExtensionAppAssembly)
                End If
                Log("Configure complete.")
                Log("---------------------------------------------")
                Log("---------------------------------------------")
            Catch ex As Exception
                Dim window As WpfUi.Templates.WindowAutoCADException = New WpfUi.Templates.WindowAutoCADException(ex)
                window.Show()
            End Try

        End Sub












        Public Sub Reload(currentIExtensionAppAssembly As Assembly)
            Try
                Log("---------------------------------------------")
                Log("---------------------------------------------")
                Log("Reload started.")
                ' If Terminated = True
                ' And Versions don't match
                ' this is the first Initalize call from any IExtensionApplication in this AutoCAD session
                Dim wasLastReloaderStateTerminated As Boolean = GetTerminatedFlag()
                If wasLastReloaderStateTerminated = True Then
                    'Not NetReloader.GetVersion().Equals(iExtensionAppVersion) Then
                    Dim dllPath As String = NetUtils.AssemblyUtils.GetFileLocationFromCodeBase(currentIExtensionAppAssembly)
                    ReloadAllDllsFoundInSameFolder(dllPath)
                Else
                    SetReloadedValues(currentIExtensionAppAssembly)
                End If

                If String.IsNullOrEmpty(GetIExtensionApplicationClassName()) Then
                    SetIExtensionApplicationClassNameFromAssembly(currentIExtensionAppAssembly)
                End If
                Log("Reload complete.")
                Log("---------------------------------------------")
                Log("---------------------------------------------")
            Catch ex As Exception
                Dim window As WpfUi.Templates.WindowAutoCADException = New WpfUi.Templates.WindowAutoCADException(ex)
                window.Show()
            End Try
        End Sub

        'Called from DllReloadClickCommandHandler
        Public Sub ReloadDll(doc As Document, iExtensionAppAssembly As Assembly, dllPath As String)

            If (dllPath IsNot Nothing) Then

                Try
                    _document = doc
                    Log("---------------------------------------------")
                    Log("---------------------------------------------")
                    Log("Dll reload started.")
                    WriteIniPathToDocEditor()
                    'Remove all commands from iExtensionAppAssembly
                    CommandRemover.RemoveAllCommandsFromiExtensionAppAssembly(doc, iExtensionAppAssembly, dllPath)
                    'RemoveAllCommandsFromAllAssembliesInAppDomain(doc, dllPath)
                    Dim tuple As Tuple(Of Assembly, String) = ReloadAllDllsFoundInSameFolder(dllPath)
                    Dim appAssembly As Assembly = tuple.Item1
                    Dim copiedMainDll As String = tuple.Item2

                    If DependencyValues.OriginalAppDirectory Is Nothing Then
                        DependencyValues.OriginalAppDirectory = IO.Path.GetDirectoryName(Assembly.GetExecutingAssembly.Location)
                    End If
                    Dim originalDirectory As String = DependencyValues.OriginalAppDirectory
                    Log(String.Format("Original app directory: {0}", originalDirectory))
                    Log(String.Format("Dll reload path: {0}", copiedMainDll))
                    Dim types As Type() = cadwiki.NetUtils.AssemblyUtils.GetTypesSafely(appAssembly)
                    WriteIniPathToDocEditor()
                    Log("Dll reload complete.")
                    Log("---------------------------------------------")
                    Log("---------------------------------------------")
                Catch ex As Exception
                    Log("Exception" + ex.Message)
                End Try
            End If
        End Sub






        Private Function ReloadAllDllsFoundInSameFolder(dllPath As String) As Tuple(Of Assembly, String)
            Try
                Log(String.Format("Reload count {0}.", DependencyValues.ReloadCount))
                Dim newCount As Integer = DependencyValues.ReloadCount + 1
                Dim dllRepository As String = Path.GetDirectoryName(dllPath)
                Dim now As DateTime = DateTime.Now
                _tempFolder = GetNewReloadFolder(newCount, now)
                IO.Directory.CreateDirectory(_tempFolder)
                Log("Created temp folder to copy dlls to for reloading: " + _tempFolder)
                Dim tempDlls As List(Of String) = CopyAllDllsToTempFolder(dllPath, _tempFolder)
                Dim tuple As Tuple(Of Assembly, String) = ReloadAll(tempDlls, newCount)
                Return tuple
            Catch ex As Exception
                Dim window As WpfUi.Templates.WindowAutoCADException = New WpfUi.Templates.WindowAutoCADException(ex)
                window.Show()
                Return Nothing
            End Try
        End Function

        Private Function CopyAllDllsToTempFolder(dllPath As String, tempFolder As String) As List(Of String)
            Dim tempDlls As New List(Of String)
            For Each dllFilePath As String In Directory.GetFiles(Path.GetDirectoryName(dllPath), "*.dll")
                Dim dllName As String = Path.GetFileName(dllFilePath)
                Dim tempFolderFilePath As String = tempFolder + "\" + dllName
                IO.File.Copy(dllFilePath, tempFolderFilePath)
                Log("Copied: " + tempFolderFilePath)
                tempDlls.Add(tempFolderFilePath)
            Next
            Return tempDlls
        End Function

        Private Function ReloadAll(tempDlls As List(Of String), reloadCount As Integer) As Tuple(Of Assembly, String)
            Dim assemblies As Assembly() = AppDomain.CurrentDomain.GetAssemblies()
            Dim appAssembly As Assembly = Nothing
            Dim appAssemblyPath As String = ""
            Log("Looking for dlls to reload.")
            Log("Skipping these dlls: ")
            For Each dllToSkip As String In DependencyValues.DllsToSkip
                Log(dllToSkip)
            Next
            appAssemblyPath = AddDllsToReloadToList(tempDlls, assemblies, appAssemblyPath)
            If String.IsNullOrEmpty(appAssemblyPath) Then
                Dim errorMessage As String = "Unable to locate the Assembly whose name contains: " + DependencyValues.IExtensionApplicationClassName
                Log(errorMessage)
            Else
                'Add appAssembly as the last Item of the list, to ensure all other dlls are loaded before
                AddDllToReload(appAssemblyPath)
            End If
            WriteInfoAboutDllsToReload()
            'Remove all commands that will be reloaded into the app domain
            RemoveAllCommandsFromAnyAssemblyThatWillBeReloaded()
            ReloadDllsIntoAppDomain()
            Return New Tuple(Of Assembly, String)(appAssembly, appAssemblyPath)
        End Function


        Private Sub RemoveAllCommandsFromAnyAssemblyThatWillBeReloaded()
            Dim assemblies As Assembly() = AppDomain.CurrentDomain.GetAssemblies()
            For Each dllFilePath As String In DependencyValues.DLLsToReload
                Dim dllName As String = IO.Path.GetFileName(dllFilePath)
                Dim assemblyName As String = IO.Path.GetFileNameWithoutExtension(dllName)
                Dim newestAssemblyWithNameInAppDomain As Assembly =
                        AcadAssemblyUtils.GetNewestAssembly(assemblies, assemblyName, Nothing)
                'Remove any commands that need to be overwritten latter
                If Not newestAssemblyWithNameInAppDomain Is Nothing Then
                    CommandRemover.RemoveAllCommandsFromiExtensionAppAssembly(_document, newestAssemblyWithNameInAppDomain, DependencyValues.OriginalAppDirectory)
                End If

            Next
        End Sub

        Private Function AddDllsToReloadToList(tempDlls As List(Of String), assemblies() As Assembly, appAssemblyPath As String) As String
            DependencyValues.DLLsToReload.Clear()
            For Each tempDll As String In tempDlls
                If File.Exists(tempDll) Then
                    Dim dllName As String = IO.Path.GetFileName(tempDll)
                    Dim assemblyName As String = IO.Path.GetFileNameWithoutExtension(tempDll)
                    Dim newestAssemblyWithNameInAppDomain As Assembly =
                        AcadAssemblyUtils.GetNewestAssembly(assemblies, assemblyName, Nothing)
                    ' If there is not an existing assembly like this in the app domain
                    ' add to list
                    If newestAssemblyWithNameInAppDomain Is Nothing Then
                        AddDllToReload(tempDll)
                    Else
                        'skip all cadwiki dlls
                        If dllName.Contains("cadwiki.") Then
                            Log("Skipped cadwiki dll: " + dllName)
                            Continue For
                        End If
                        ' Else do version checks
                        Dim newestVersionInAppDomain As String =
                        AcadAssemblyUtils.GetAssemblyVersionFromFullName(newestAssemblyWithNameInAppDomain.FullName)
                        Dim copiedFvi As FileVersionInfo = FileVersionInfo.GetVersionInfo(tempDll)
                        Dim copiedVersion As String = copiedFvi.FileVersion
                        Dim isCopiedDllNewer As Integer = AcadAssemblyUtils.CompareFileVersion(copiedVersion, CStr(newestVersionInAppDomain))
                        ' For dlls that already exist in the app domain, only reload if the copy has a newer file version
                        ' which result in a 1 comparision between file versions
                        If isCopiedDllNewer = 1 Then
                            ' If dll name contains project name, don't store to list
                            If assemblyName.Contains(DependencyValues.IExtensionApplicationClassName) Then
                                Log("Skipped iExtension app dll")
                                appAssemblyPath = tempDll
                            Else
                                Log(dllName + " added to list. App domain version:" + newestVersionInAppDomain + ", copied version:" + copiedVersion)
                                'Else store to list
                                AddDllToReload(tempDll)
                            End If
                        Else
                            Log(dllName + " skipped. App domain version:" + newestVersionInAppDomain + ", copied version:" + copiedVersion)
                        End If
                    End If

                Else
                End If
            Next

            Return appAssemblyPath
        End Function

        Private Sub ReloadDllsIntoAppDomain()
            Dim assemblyWithIExtensionApp As Assembly = Nothing
            For Each dllPath As String In DependencyValues.DLLsToReload
                Dim assemblyBytes As Byte() = Nothing
                Try
                    assemblyBytes = System.IO.File.ReadAllBytes(dllPath)
                Catch ex As Exception
                    Log("Error reading assembly to byte array: " + dllPath)
                    Log("Exception: " + ex.Message)
                End Try
                Try
                    If dllPath.Contains(DependencyValues.IExtensionApplicationClassName) Then
                        'Update Reloader values
                        DependencyValues.ReloadCount += 1
                        DependencyValues.Terminated = False
                        WriteDependecyValuesToIni(DependencyValues)
                        ' Upon loading the assemblyBytes from the IExtensionApplication class, the App.Initialize() method will be called
                        assemblyWithIExtensionApp = AppDomain.CurrentDomain.Load(assemblyBytes)
                        Log("Reloaded iExtensionAppAssembly dll: " + dllPath)
                        SetReloadedValues(assemblyWithIExtensionApp)
                        WriteDependecyValuesToIni(DependencyValues)
                    Else
                        Dim reloadedAssembly As Assembly = AppDomain.CurrentDomain.Load(assemblyBytes)
                        Log("Reloaded dll: " + dllPath)
                    End If
                Catch ex As Exception
                    Log("Error loading assembly: " + dllPath)
                    Log("Exception: " + ex.Message)
                End Try
            Next
            Dim currentTypes As Type() = cadwiki.NetUtils.AssemblyUtils.GetTypesSafely(assemblyWithIExtensionApp)
            ' Create reference to the IExtensionApplication object
            Dim currentAppObject As Object = AcadAssemblyUtils.GetAppObjectSafely(currentTypes)
            'currentAppObject.Initialize()
        End Sub



















        Public Shadows Sub Log(message As String)
            Dim mode As AutoCADAppDomainDllReloader.LogMode = GetLogMode()
            Select Case mode.Equals(AutoCADAppDomainDllReloader.LogMode.Off)
                Case mode.Equals(AutoCADAppDomainDllReloader.LogMode.Text)
                    LogToTextFile(message)
                Case mode.Equals(AutoCADAppDomainDllReloader.LogMode.AcadDocEditor)
                    LogToTextFile(message)
                    LogToEditor(message)
            End Select
        End Sub

        Private Sub LogToTextFile(message As String)
            If (TextFileLog Is Nothing) Then
                TextFileLog = New NetUtils.TextFileLog()
                TextFileLog.Write(message)
            Else
                TextFileLog.Write(message)
            End If
        End Sub

        Private Sub LogToEditor(message As String)
            If (_document IsNot Nothing) Then
                _document.Editor.WriteMessage(vbLf & message)
                _document.Editor.WriteMessage(vbLf)
            End If
        End Sub

        Public Sub LogException(ex As Exception)
            Dim mode As AutoCADAppDomainDllReloader.LogMode = GetLogMode()
            Select Case mode.Equals(AutoCADAppDomainDllReloader.LogMode.Off)
                Case mode.Equals(AutoCADAppDomainDllReloader.LogMode.Text)
                    LogExceptionToTextFile(ex)
                Case mode.Equals(AutoCADAppDomainDllReloader.LogMode.AcadDocEditor)
                    LogExceptionToTextFile(ex)
                    LogExceptionToEditor(ex)
            End Select
        End Sub

        Private Sub LogExceptionToTextFile(ex As Exception)
            If (TextFileLog Is Nothing) Then
                TextFileLog = New NetUtils.TextFileLog()
                TextFileLog.Exception(ex)
            Else
                TextFileLog.Exception(ex)
            End If
        End Sub

        Private Sub LogExceptionToEditor(ex As Exception)
            If (_document Is Nothing) Then
                _document = Autodesk.AutoCAD.ApplicationServices.Application.DocumentManager.MdiActiveDocument
            End If

            Dim ed As Editor = _document.Editor
            Dim list As List(Of String) = NetUtils.Exceptions.GetPrettyStringList(ex)
            For Each str As String In list
                ed.WriteMessage(Environment.NewLine.ToString() & str)
            Next

        End Sub

    End Class

End Namespace

File: cadwiki-nuget/cadwiki.DllReloader/AutoCAD/AutoCADAppDomainDllReloader.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Reflection;
using System.Windows.Controls;
using Autodesk.AutoCAD.ApplicationServices;
using Microsoft.VisualBasic;

namespace cadwiki.DllReloader.AutoCAD
{
    public class AutoCADAppDomainDllReloader : AutodeskAppDomainReloader
    {

        private Document _document;
        private string _tempFolder;

        public AutoCADAppDomainDllReloader()
        {
            _tempFolder = CadwikiTempFolder;
        }


        public new LogMode GetLogMode()
        {
            return DependencyValues.LogMode;
        }

        public void Configure(Assembly currentIExtensionAppAssembly)
        {
            try
            {
                Log("---------------------------------------------");
                Log("---------------------------------------------");
                Log("Configure started.");
                ReadDependecyValuesFromIni();
                // If Terminated = True
                // And Versions don't match
                // this is the first Initalize call from any IExtensionApplication in this AutoCAD session
                bool wasLastReloaderStateTerminated = GetTerminatedFlag();
                if (wasLastReloaderStateTerminated == true)
                {
                    // Not NetReloader.GetVersion().Equals(iExtensionAppVersion) Then
                    SetReloadCount(0);
                    SetInitialValues(currentIExtensionAppAssembly);
                }

                if (string.IsNullOrEmpty(GetIExtensionApplicationClassName()))
                {
                    SetIExtensionApplicationClassNameFromAssembly(currentIExtensionAppAssembly);
                }
                Log("Configure complete.");
                Log("---------------------------------------------");
                Log("---------------------------------------------");
            }
            catch (Exception ex)
            {
                var window = new WpfUi.Templates.WindowAutoCADException(ex);
                window.Show();
            }

        }












        public void Reload(Assembly currentIExtensionAppAssembly)
        {
            try
            {
                Log("---------------------------------------------");
                Log("---------------------------------------------");
                Log("Reload started.");
                // If Terminated = True
                // And Versions don't match
                // this is the first Initalize call from any IExtensionApplication in this AutoCAD session
                bool wasLastReloaderStateTerminated = GetTerminatedFlag();
                if (wasLastReloaderStateTerminated == true)
                {
                    // Not NetReloader.GetVersion().Equals(iExtensionAppVersion) Then
                    string dllPath = NetUtils.AssemblyUtils.GetFileLocationFromCodeBase(currentIExtensionAppAssembly);
                    ReloadAllDllsFoundInSameFolder(dllPath);
                }
                else
                {
                    SetReloadedValues(currentIExtensionAppAssembly);
                }

                if (string.IsNullOrEmpty(GetIExtensionApplicationClassName()))
                {
                    SetIExtensionApplicationClassNameFromAssembly(currentIExtensionAppAssembly);
                }
                Log("Reload complete.");
                Log("---------------------------------------------");
                Log("---------------------------------------------");
            }
            catch (Exception ex)
            {
                var window = new WpfUi.Templates.WindowAutoCADException(ex);
                window.Show();
            }
        }

        // -------------------------------------------------------------------------
        // Phase 2 — staged-folder reload entry point
        // -------------------------------------------------------------------------

        /// <summary>
        /// Thin wrapper that delegates to the existing <see cref="ReloadAll"/> pipeline
        /// using a pre-staged folder produced by <c>cadwiki.PluginReloadService.StagingCopier</c>.
        ///
        /// <para>
        ///   Unlike <see cref="ReloadDll"/>, this method does <em>not</em> require a
        ///   live <see cref="Assembly"/> reference for the plugin — making it safe for
        ///   <em>external</em> plugins that were not loaded by the cadwiki assemblies.
        /// </para>
        ///
        /// <para>
        ///   The staged folder must already contain:
        ///   <list type="bullet">
        ///     <item>Version-rewritten copies of the managed plugin DLLs.</item>
        ///     <item>Verbatim copies of all other DLLs and supporting files.</item>
        ///   </list>
        ///   These are produced by <c>cadwiki.PluginReloadService.StagingCopier.StagePlugin()</c>.
        /// </para>
        /// </summary>
        /// <param name="doc">
        ///   Active AutoCAD document. Used for command removal and editor log.
        /// </param>
        /// <param name="stagedFolderPath">
        ///   Full path to the staging folder.
        ///   Example: <c>%TEMP%\cadwiki.PluginStaging\MyPlugin\20231015--14_22_30--Build-3\</c>
        /// </param>
        /// <param name="mainDllName">
        ///   Filename of the main plugin DLL inside the staged folder
        ///   (e.g. <c>MyPlugin.dll</c>).
        /// </param>
        public void ReloadFromStagedFolder(Document doc, string stagedFolderPath, string mainDllName)
        {
            if (stagedFolderPath is null)
                throw new ArgumentNullException(nameof(stagedFolderPath));
            if (mainDllName is null)
                throw new ArgumentNullException(nameof(mainDllName));

            try
            {
                _document = doc;
                Log("---------------------------------------------");
                Log("---------------------------------------------");
                Log($"ReloadFromStagedFolder started.");
                Log($"Staged folder : {stagedFolderPath}");
                Log($"Main DLL      : {mainDllName}");

                WriteIniPathToDocEditor();

                // Build the list of DLLs in the staged folder (same shape as
                // CopyAllDllsToTempFolder output, but already at their final paths).
                var stagedDlls = new System.Collections.Generic.List<string>();
                foreach (string dllFilePath in Directory.GetFiles(stagedFolderPath, "*.dll"))
                    stagedDlls.Add(dllFilePath);

                Log($"Found {stagedDlls.Count} DLL(s) in staged folder.");

                // Ensure IExtensionApplicationClassName is set so ReloadAll can
                // identify the main plugin assembly.
                if (string.IsNullOrEmpty(DependencyValues.IExtensionApplicationClassName))
                {
                    string mainName = Path.GetFileNameWithoutExtension(mainDllName);
                    SetIExtensionApplicationClassName(mainName);
                    Log($"IExtensionApplicationClassName set to: {mainName}");
                }

                // Determine build number (reuse counter from dependency values)
                int newCount = DependencyValues.ReloadCount + 1;

                // Store the staged folder as the temp folder so log + ini paths align
                _tempFolder = stagedFolderPath;

                // Remove commands from the existing version of the plugin before reload
                TryRemoveAllCommandsExternal(doc, stagedFolderPath, mainDllName);

                // Delegate to the shared ReloadAll pipeline — no code duplication
                var tuple = ReloadAll(stagedDlls, newCount);

                if (DependencyValues.OriginalAppDirectory is null)
                    DependencyValues.OriginalAppDirectory = stagedFolderPath;

                Log($"ReloadFromStagedFolder complete. Main assembly: {tuple?.Item2 ?? "(none)"}");
                Log("---------------------------------------------");
                Log("---------------------------------------------");
            }
            catch (Exception ex)
            {
                Log($"ReloadFromStagedFolder exception: {ex.Message}");
                Log($"StackTrace: {ex.StackTrace}");
            }
        }

        /// <summary>
        /// Attempts to remove commands for an external plugin by scanning the
        /// current AppDomain for an assembly whose name matches <paramref name="mainDllName"/>.
        /// Swallows all exceptions — command removal is best-effort.
        /// </summary>
        private void TryRemoveAllCommandsExternal(Document doc, string stagedFolderPath, string mainDllName)
        {
            try
            {
                string assemblyName = Path.GetFileNameWithoutExtension(mainDllName);
                var assemblies = AppDomain.CurrentDomain.GetAssemblies();
                var existing = AcadAssemblyUtils.GetNewestAssembly(assemblies, assemblyName, null);
                if (existing != null)
                {
                    Log($"Removing commands from existing assembly: {existing.FullName}");
                    TryRemoveAllCommands(doc, existing, stagedFolderPath);
                }
                else
                {
                    Log($"No existing assembly named '{assemblyName}' found in AppDomain; skipping command removal.");
                }
            }
            catch (Exception ex)
            {
                Log($"TryRemoveAllCommandsExternal exception (ignored): {ex.Message}");
            }
        }

        // -------------------------------------------------------------------------
        // End of staged-folder reload — existing methods follow
        // -------------------------------------------------------------------------

        // Called from DllReloadClickCommandHandler
        public void ReloadDll(Document doc, Assembly iExtensionAppAssembly, string dllPath)
        {

            if (dllPath is not null)
            {

                try
                {
                    _document = doc;
                    Log("---------------------------------------------");
                    Log("---------------------------------------------");
                    Log("Dll reload started.");
                    WriteIniPathToDocEditor();
                    // Remove all commands from iExtensionAppAssembly
                    TryRemoveAllCommands(doc, iExtensionAppAssembly, dllPath);
                    // RemoveAllCommandsFromAllAssembliesInAppDomain(doc, dllPath)
                    var tuple = ReloadAllDllsFoundInSameFolder(dllPath);
                    var appAssembly = tuple.Item1;
                    string copiedMainDll = tuple.Item2;

                    if (DependencyValues.OriginalAppDirectory is null)
                    {
                        DependencyValues.OriginalAppDirectory = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
                    }
                    string originalDirectory = DependencyValues.OriginalAppDirectory;
                    Log(string.Format("Original app directory: {0}", originalDirectory));
                    Log(string.Format("Dll reload path: {0}", copiedMainDll));
                    Type[] types = NetUtils.AssemblyUtils.GetTypesSafely(appAssembly);
                    WriteIniPathToDocEditor();
                    Log("Dll reload complete.");
                    Log("---------------------------------------------");
                    Log("---------------------------------------------");
                }
                catch (Exception ex)
                {
                    Log("Exception" + ex.Message);
                }
            }
        }

        private void TryRemoveAllCommands(Document doc, Assembly iExtensionAppAssembly, string dllPath)
        {
            try
            {
                CommandRemover.RemoveAllCommandsFromiExtensionAppAssembly(doc, iExtensionAppAssembly, dllPath);
            }
            catch (Exception ex)
            {
                Log("Exception" + ex.Message);
            }
        }

        private Tuple<Assembly, string> ReloadAllDllsFoundInSameFolder(string dllPath)
        {
            try
            {
                Log(string.Format("Reload count {0}.", DependencyValues.ReloadCount));
                int newCount = DependencyValues.ReloadCount + 1;
                string dllRepository = Path.GetDirectoryName(dllPath);
                var now = DateTime.Now;
                _tempFolder = GetNewReloadFolder(newCount, now);
                Directory.CreateDirectory(_tempFolder);
                Log("Created temp folder to copy dlls to for reloading: " + _tempFolder);
                var tempDlls = CopyAllDllsToTempFolder(dllPath, _tempFolder);
                var tuple = ReloadAll(tempDlls, newCount);
                return tuple;
            }
            catch (Exception ex)
            {
                var window = new WpfUi.Templates.WindowAutoCADException(ex);
                window.Show();
                return null;
            }
        }

        private List<string> CopyAllDllsToTempFolder(string dllPath, string tempFolder)
        {
            var tempDlls = new List<string>();
            foreach (string dllFilePath in Directory.GetFiles(Path.GetDirectoryName(dllPath), "*.dll"))
            {
                string dllName = Path.GetFileName(dllFilePath);
                string tempFolderFilePath = tempFolder + @"\" + dllName;
                File.Copy(dllFilePath, tempFolderFilePath);
                Log("Copied: " + tempFolderFilePath);
                tempDlls.Add(tempFolderFilePath);
            }
            return tempDlls;
        }

        private Tuple<Assembly, string> ReloadAll(List<string> tempDlls, int reloadCount)
        {
            Assembly[] assemblies = AppDomain.CurrentDomain.GetAssemblies();
            Assembly appAssembly = null;
            string appAssemblyPath = "";
            Log("Looking for dlls to reload.");
            Log("Skipping these dlls: ");
            foreach (string dllToSkip in DependencyValues.DllsToSkip)
                Log(dllToSkip);
            appAssemblyPath = AddDllsToReloadToList(tempDlls, assemblies, appAssemblyPath);
            if (string.IsNullOrEmpty(appAssemblyPath))
            {
                string errorMessage = "Unable to locate the Assembly whose name contains: " + DependencyValues.IExtensionApplicationClassName;
                Log(errorMessage);
            }
            else
            {
                // Add appAssembly as the last Item of the list, to ensure all other dlls are loaded before
                AddDllToReload(appAssemblyPath);
            }
            WriteInfoAboutDllsToReload();
            // Remove all commands that will be reloaded into the app domain
            RemoveAllCommandsFromAnyAssemblyThatWillBeReloaded();
            ReloadDllsIntoAppDomain();
            return new Tuple<Assembly, string>(appAssembly, appAssemblyPath);
        }


        private void RemoveAllCommandsFromAnyAssemblyThatWillBeReloaded()
        {
            Assembly[] assemblies = AppDomain.CurrentDomain.GetAssemblies();
            foreach (string dllFilePath in DependencyValues.DLLsToReload)
            {
                string dllName = Path.GetFileName(dllFilePath);
                string assemblyName = Path.GetFileNameWithoutExtension(dllName);
                var newestAssemblyWithNameInAppDomain = AcadAssemblyUtils.GetNewestAssembly(assemblies, assemblyName, null);
                // Remove any commands that need to be overwritten latter
                if (newestAssemblyWithNameInAppDomain is not null)
                {
                    TryRemoveAllCommands(_document, newestAssemblyWithNameInAppDomain, DependencyValues.OriginalAppDirectory);
                }

            }
        }

        private string AddDllsToReloadToList(List<string> tempDlls, Assembly[] assemblies, string appAssemblyPath)
        {
            DependencyValues.DLLsToReload.Clear();
            foreach (string tempDll in tempDlls)
            {
                if (File.Exists(tempDll))
                {
                    string dllName = Path.GetFileName(tempDll);
                    string assemblyName = Path.GetFileNameWithoutExtension(tempDll);
                    var newestAssemblyWithNameInAppDomain = AcadAssemblyUtils.GetNewestAssembly(assemblies, assemblyName, null);
                    // If there is not an existing assembly like this in the app domain
                    // add to list
                    if (newestAssemblyWithNameInAppDomain is null)
                    {
                        AddDllToReload(tempDll);
                    }
                    else
                    {
                        // skip all cadwiki dlls
                        if (SkipCadwikiDlls & dllName.Contains("cadwiki."))
                        {
                            Log("Skipped cadwiki dll: " + dllName);
                            continue;
                        }
                        // Else do version checks
                        string newestVersionInAppDomain = AcadAssemblyUtils.GetAssemblyVersionFromFullName(newestAssemblyWithNameInAppDomain.FullName);
                        var copiedFvi = FileVersionInfo.GetVersionInfo(tempDll);
                        string copiedVersion = copiedFvi.FileVersion;
                        int isCopiedDllNewer = AcadAssemblyUtils.CompareFileVersion(copiedVersion, newestVersionInAppDomain);
                        // For dlls that already exist in the app domain, only reload if the copy has a newer file version
                        // which result in a 1 comparision between file versions
                        if (isCopiedDllNewer == 1)
                        {
                            // If dll name contains project name, don't store to list
                            if (assemblyName.Contains(DependencyValues.IExtensionApplicationClassName))
                            {
                                Log("Skipped iExtension app dll");
                                appAssemblyPath = tempDll;
                            }
                            else
                            {
                                Log(dllName + " added to list. App domain version:" + newestVersionInAppDomain + ", copied version:" + copiedVersion);
                                // Else store to list
                                AddDllToReload(tempDll);
                            }
                        }
                        else
                        {
                            Log(dllName + " skipped. App domain version:" + newestVersionInAppDomain + ", copied version:" + copiedVersion);
                        }
                    }
                }

                else
                {
                }
            }

            return appAssemblyPath;
        }

        private void ReloadDllsIntoAppDomain()
        {
            Assembly assemblyWithIExtensionApp = null;
            foreach (string dllPath in DependencyValues.DLLsToReload)
            {
                byte[] assemblyBytes = null;
                try
                {
                    assemblyBytes = File.ReadAllBytes(dllPath);
                }
                catch (Exception ex)
                {
                    Log("Error reading assembly to byte array: " + dllPath);
                    Log("Exception: " + ex.Message);
                }
                try
                {
                    if (dllPath.Contains(DependencyValues.IExtensionApplicationClassName))
                    {
                        // Update Reloader values
                        DependencyValues.ReloadCount += 1;
                        DependencyValues.Terminated = false;
                        WriteDependecyValuesToIni(DependencyValues);
                        // Upon loading the assemblyBytes from the IExtensionApplication class, the App.Initialize() method will be called
                        assemblyWithIExtensionApp = AppDomain.CurrentDomain.Load(assemblyBytes);
                        Log("Reloaded iExtensionAppAssembly dll: " + dllPath);
                        SetReloadedValues(assemblyWithIExtensionApp);
                        WriteDependecyValuesToIni(DependencyValues);
                    }
                    else
                    {
                        var reloadedAssembly = AppDomain.CurrentDomain.Load(assemblyBytes);
                        Log("Reloaded dll: " + dllPath);
                    }
                }
                catch (Exception ex)
                {
                    Log("Error loading assembly: " + dllPath);
                    Log("Exception: " + ex.Message);
                }
            }
            Type[] currentTypes = NetUtils.AssemblyUtils.GetTypesSafely(assemblyWithIExtensionApp);
            // Create reference to the IExtensionApplication object
            var currentAppObject = AcadAssemblyUtils.GetAppObjectSafely(currentTypes);
            // currentAppObject.Initialize()
        }



















        public new void Log(string message)
        {
            var mode = GetLogMode();
            switch (mode.Equals(LogMode.Off))
            {
                case object _ when mode.Equals(LogMode.Text):
                    {
                        LogToTextFile(message);
                        break;
                    }
                case object _ when mode.Equals(LogMode.AcadDocEditor):
                    {
                        LogToTextFile(message);
                        LogToEditor(message);
                        break;
                    }
            }
        }

        private new void LogToTextFile(string message)
        {
            if (ReloaderLog is null)
            {
                ReloaderLog = new ReloaderLog();
                ReloaderLog.LogDir = _tempFolder;
                ReloaderLog.Write(message);
            }
            else
            {
                ReloaderLog.LogDir = _tempFolder;
                ReloaderLog.Write(message);
            }
        }

        private void LogToEditor(string message)
        {
            if (_document is not null)
            {
                _document.Editor.WriteMessage(Environment.NewLine + message);
                _document.Editor.WriteMessage(Environment.NewLine);
            }
        }

        public new void LogException(Exception ex)
        {
            var mode = GetLogMode();
            switch (mode.Equals(LogMode.Off))
            {
                case object _ when mode.Equals(LogMode.Text):
                    {
                        LogExceptionToTextFile(ex);
                        break;
                    }
                case object _ when mode.Equals(LogMode.AcadDocEditor):
                    {
                        LogExceptionToTextFile(ex);
                        LogExceptionToEditor(ex);
                        break;
                    }
            }
        }

        private new void LogExceptionToTextFile(Exception ex)
        {
            if (ReloaderLog is null)
            {
                ReloaderLog = new ReloaderLog();
                ReloaderLog.LogDir = _tempFolder;
                ReloaderLog.Exception(ex);
            }
            else
            {
                ReloaderLog.LogDir = _tempFolder;
                ReloaderLog.Exception(ex);
            }
        }

        private void LogExceptionToEditor(Exception ex)
        {
            if (_document is null)
            {
                _document = Autodesk.AutoCAD.ApplicationServices.Core.Application.DocumentManager.MdiActiveDocument;
            }

            var ed = _document.Editor;
            var list = NetUtils.Exceptions.GetPrettyStringList(ex);
            foreach (string str in list)
                ed.WriteMessage(Environment.NewLine.ToString() + str);

        }




    }

}